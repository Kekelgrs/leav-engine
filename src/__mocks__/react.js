const react = require('react');
module.exports = {...react, memo: x => x};

export default undefined;
