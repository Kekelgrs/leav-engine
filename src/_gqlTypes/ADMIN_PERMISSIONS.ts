/* tslint:disable */
/* eslint-disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: ADMIN_PERMISSIONS
// ====================================================

export interface ADMIN_PERMISSIONS_adminPermissions {
    admin_access_attributes: boolean;
    admin_access_libraries: boolean;
    admin_access_permissions: boolean;
    admin_access_trees: boolean;
    admin_create_attribute: boolean;
    admin_create_library: boolean;
    admin_create_tree: boolean;
    admin_delete_attribute: boolean;
    admin_delete_library: boolean;
    admin_delete_tree: boolean;
    admin_edit_attribute: boolean;
    admin_edit_library: boolean;
    admin_edit_permission: boolean;
    admin_edit_tree: boolean;
}

export interface ADMIN_PERMISSIONS {
    adminPermissions: ADMIN_PERMISSIONS_adminPermissions | null;
}
