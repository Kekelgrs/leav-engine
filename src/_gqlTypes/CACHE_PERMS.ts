/* tslint:disable */
/* eslint-disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL mutation operation: CACHE_PERMS
// ====================================================

export interface CACHE_PERMS {
    cachePerms: any | null;
}

export interface CACHE_PERMSVariables {
    perms: any;
}
