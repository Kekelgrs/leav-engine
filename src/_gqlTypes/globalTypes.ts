/* tslint:disable */
/* eslint-disable */
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

export enum ActionIOTypes {
    boolean = 'boolean',
    number = 'number',
    object = 'object',
    string = 'string'
}

export enum AttributeFormat {
    boolean = 'boolean',
    date = 'date',
    encrypted = 'encrypted',
    extended = 'extended',
    numeric = 'numeric',
    text = 'text'
}

export enum AttributeType {
    advanced = 'advanced',
    advanced_link = 'advanced_link',
    simple = 'simple',
    simple_link = 'simple_link',
    tree = 'tree'
}

export enum AvailableActionsName {
    encrypt = 'encrypt',
    formatDate = 'formatDate',
    formatNumber = 'formatNumber',
    maskValue = 'maskValue',
    parseJSON = 'parseJSON',
    toBoolean = 'toBoolean',
    toJSON = 'toJSON',
    toNumber = 'toNumber',
    toString = 'toString',
    validateFormat = 'validateFormat',
    validateRegex = 'validateRegex'
}

export enum AvailableLanguage {
    en = 'en',
    fr = 'fr'
}

export enum PermissionTypes {
    admin = 'admin',
    attribute = 'attribute',
    library = 'library',
    record = 'record'
}

export enum PermissionsActions {
    access = 'access',
    access_attribute = 'access_attribute',
    admin_access_attributes = 'admin_access_attributes',
    admin_access_libraries = 'admin_access_libraries',
    admin_access_navigator = 'admin_access_navigator',
    admin_access_permissions = 'admin_access_permissions',
    admin_access_trees = 'admin_access_trees',
    admin_create_attribute = 'admin_create_attribute',
    admin_create_library = 'admin_create_library',
    admin_create_tree = 'admin_create_tree',
    admin_delete_attribute = 'admin_delete_attribute',
    admin_delete_library = 'admin_delete_library',
    admin_delete_tree = 'admin_delete_tree',
    admin_edit_attribute = 'admin_edit_attribute',
    admin_edit_library = 'admin_edit_library',
    admin_edit_permission = 'admin_edit_permission',
    admin_edit_tree = 'admin_edit_tree',
    create = 'create',
    create_value = 'create_value',
    delete = 'delete',
    delete_value = 'delete_value',
    edit = 'edit',
    edit_value = 'edit_value'
}

export enum PermissionsRelation {
    and = 'and',
    or = 'or'
}

export enum ValueVersionMode {
    simple = 'simple',
    smart = 'smart'
}

export interface ActionConfigurationInput {
    name: AvailableActionsName;
    params?: (ActionConfigurationParamInput | null)[] | null;
}

export interface ActionConfigurationParamInput {
    name: string;
    value: string;
}

export interface ActionsListConfigurationInput {
    saveValue?: (ActionConfigurationInput | null)[] | null;
    getValue?: (ActionConfigurationInput | null)[] | null;
    deleteValue?: (ActionConfigurationInput | null)[] | null;
}

export interface AttributeInput {
    id: string;
    type?: AttributeType | null;
    format?: AttributeFormat | null;
    label?: SystemTranslationInput | null;
    linked_library?: string | null;
    linked_tree?: string | null;
    embedded_fields?: (EmbeddedAttributeInput | null)[] | null;
    actions_list?: ActionsListConfigurationInput | null;
    permissions_conf?: Treepermissions_confInput | null;
    multiple_values?: boolean | null;
    versions_conf?: ValuesVersionsConfInput | null;
}

export interface EmbeddedAttributeInput {
    id: string;
    format?: AttributeFormat | null;
    label?: SystemTranslationInput | null;
    validation_regex?: string | null;
    embedded_fields?: (EmbeddedAttributeInput | null)[] | null;
}

export interface LibraryInput {
    id: string;
    label?: SystemTranslationInput | null;
    attributes?: string[] | null;
    permissions_conf?: Treepermissions_confInput | null;
    recordIdentityConf?: RecordIdentityConfInput | null;
}

export interface PermissionActionInput {
    name: PermissionsActions;
    allowed?: boolean | null;
}

/**
 * If users group is not specified, permission will be saved at root level.
 * If saving a tree-based permission (record or attribute) and tree target's id is not specified,
 * permission will be saved at root level for any element of the tree.
 */
export interface PermissionInput {
    type: PermissionTypes;
    applyTo?: string | null;
    usersGroup?: string | null;
    actions: PermissionActionInput[];
    permissionTreeTarget?: PermissionsTreeTargetInput | null;
}

/**
 * Element on which we want to retrieve record or attribute permission. Record ID is mandatory,
 * attributeId is only required for attribute permission
 */
export interface PermissionTarget {
    attributeId?: string | null;
    recordId: string;
}

/**
 * If id and library are not specified, permission will apply to tree root
 */
export interface PermissionsTreeTargetInput {
    tree: string;
    library?: string | null;
    id?: string | null;
}

export interface RecordIdentityConfInput {
    label?: string | null;
    color?: string | null;
    preview?: string | null;
}

export interface SystemTranslationInput {
    fr: string;
    en?: string | null;
}

export interface TreeElementInput {
    id: string;
    library: string;
}

export interface TreeInput {
    id: string;
    libraries?: string[] | null;
    label?: SystemTranslationInput | null;
}

export interface Treepermissions_confInput {
    permissionTreeAttributes: string[];
    relation: PermissionsRelation;
}

export interface ValueBatchInput {
    attribute?: string | null;
    id_value?: string | null;
    value?: string | null;
}

export interface ValueVersionInput {
    name: string;
    value: TreeElementInput;
}

export interface ValuesVersionsConfInput {
    versionable: boolean;
    mode?: ValueVersionMode | null;
    trees?: string[] | null;
}

//==============================================================
// END Enums and Input Objects
//==============================================================
