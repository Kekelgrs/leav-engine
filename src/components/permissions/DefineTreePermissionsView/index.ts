import DefineTreePermissionsView from './DefineTreePermissionsView';
export default DefineTreePermissionsView;
