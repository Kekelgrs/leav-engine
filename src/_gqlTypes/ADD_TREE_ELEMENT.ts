/* tslint:disable */
/* eslint-disable */
// This file was automatically generated and should not be edited.

import {TreeElementInput} from './globalTypes';

// ====================================================
// GraphQL mutation operation: ADD_TREE_ELEMENT
// ====================================================

export interface ADD_TREE_ELEMENT_treeAddElement {
    id: string | null;
    library: string | null;
}

export interface ADD_TREE_ELEMENT {
    treeAddElement: ADD_TREE_ELEMENT_treeAddElement;
}

export interface ADD_TREE_ELEMENTVariables {
    treeId: string;
    element: TreeElementInput;
    parent?: TreeElementInput | null;
}
