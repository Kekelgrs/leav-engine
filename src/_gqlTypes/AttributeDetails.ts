/* tslint:disable */
/* eslint-disable */
// This file was automatically generated and should not be edited.

import {AttributeType, AttributeFormat, PermissionsRelation, ValueVersionMode} from './globalTypes';

// ====================================================
// GraphQL fragment: AttributeDetails
// ====================================================

export interface AttributeDetails_permissionsConf_permissionTreeAttributes {
    id: string;
    linked_tree: string | null;
    label: any | null;
}

export interface AttributeDetails_permissionsConf {
    permissionTreeAttributes: AttributeDetails_permissionsConf_permissionTreeAttributes[];
    relation: PermissionsRelation;
}

export interface AttributeDetails_versionsConf {
    versionable: boolean;
    mode: ValueVersionMode | null;
    trees: string[] | null;
}

export interface AttributeDetails {
    id: string;
    type: AttributeType;
    format: AttributeFormat | null;
    system: boolean;
    label: any | null;
    linked_library: string | null;
    linked_tree: string | null;
    multipleValues: boolean;
    permissionsConf: AttributeDetails_permissionsConf | null;
    versionsConf: AttributeDetails_versionsConf | null;
}
