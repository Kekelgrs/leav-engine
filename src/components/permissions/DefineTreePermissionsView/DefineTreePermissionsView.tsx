import * as React from 'react';
import {WithNamespaces, withNamespaces} from 'react-i18next';
import {NodeData, TreeNode} from 'react-sortable-tree';
import ColumnsDisplay from 'src/components/shared/ColumnsDisplay';
import {getTreeNodeKey} from 'src/utils/utils';
import {GET_LIBRARIES_libraries_permissionsConf_permissionTreeAttributes} from 'src/_gqlTypes/GET_LIBRARIES';
import {PermissionsActions, PermissionTypes} from 'src/_gqlTypes/globalTypes';
import DefinePermissionsViewLoadTree from '../DefinePermissionsViewLoadTree';
import EditPermissions from '../EditPermissions';

interface IDefineTreePermissionsViewProps extends WithNamespaces {
    treeAttribute: GET_LIBRARIES_libraries_permissionsConf_permissionTreeAttributes;
    permissionType: PermissionTypes;
    applyTo: string;
}

function DefineTreePermissionsView({
    treeAttribute: tree,
    permissionType,
    applyTo
}: IDefineTreePermissionsViewProps): JSX.Element {
    const usersGroupsTreeId = 'users_groups';
    const [selectedTreeNode, setSelectedTreeNode] = React.useState<NodeData | null>(null);
    const [selectedGroupNode, setSelectedGroupNode] = React.useState<NodeData | null>(null);
    const _selectTreeNode = (nodeData: NodeData) =>
        setSelectedTreeNode(
            getTreeNodeKey(nodeData) !== getTreeNodeKey(selectedTreeNode as TreeNode) ? nodeData : null
        );

    const _selectGroupNode = (nodeData: NodeData) =>
        setSelectedGroupNode(
            getTreeNodeKey(nodeData) !== getTreeNodeKey(selectedGroupNode as TreeNode) ? nodeData : null
        );

    if (!tree.linked_tree) {
        return <p>Cannot find tree</p>;
    }

    const cols = [
        <DefinePermissionsViewLoadTree
            key="perm_tree"
            treeId={tree.linked_tree}
            onClick={_selectTreeNode}
            selectedNode={selectedTreeNode}
        />
    ];

    if (selectedTreeNode) {
        cols.push(
            <DefinePermissionsViewLoadTree
                treeId={usersGroupsTreeId}
                onClick={_selectGroupNode}
                selectedNode={selectedGroupNode}
            />
        );

        if (selectedGroupNode) {
            cols.push(
                <EditPermissions
                    permParams={{
                        type: permissionType,
                        applyTo,
                        usersGroup: selectedGroupNode.node.id,
                        actions: [
                            PermissionsActions.access,
                            PermissionsActions.create,
                            PermissionsActions.edit,
                            PermissionsActions.delete
                        ],
                        permissionTreeTarget: {
                            tree: tree.linked_tree,
                            id: selectedTreeNode.node.id,
                            library: selectedTreeNode.node.library.id
                        }
                    }}
                />
            );
        }
    }

    return <ColumnsDisplay columnsNumber={3} columnsContent={cols} />;
}

export default withNamespaces()(DefineTreePermissionsView);
