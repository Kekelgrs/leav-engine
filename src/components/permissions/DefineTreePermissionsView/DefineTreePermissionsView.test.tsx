import {render} from 'enzyme';
import * as React from 'react';
import {MockedProvider} from 'react-apollo/test-utils';
import {PermissionTypes} from 'src/_gqlTypes/globalTypes';
import DefineTreePermissionsView from './DefineTreePermissionsView';

describe('DefineTreePermissionsView', () => {
    test('Snapshot test', async () => {
        const onSave = jest.fn();
        const mockTree = {
            id: 'test_tree_attr',
            linked_tree: 'test_tree',
            label: {fr: 'Test'}
        };

        const comp = render(
            <MockedProvider>
                <DefineTreePermissionsView
                    treeAttribute={mockTree}
                    onSavePermissions={onSave}
                    permissionType={PermissionTypes.record}
                    applyTo="test_lib"
                />
            </MockedProvider>
        );

        expect(comp).toMatchSnapshot();
    });
});
