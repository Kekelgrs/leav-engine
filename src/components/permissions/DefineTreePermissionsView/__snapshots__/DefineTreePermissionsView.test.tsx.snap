// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefineTreePermissionsView Snapshot test 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "style": "display:flex;flex-direction:row;justify-content:flex-start",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "style": "position:relative;padding-left:1em;width:33%;border-right:1px solid #999999",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "ui active text centered inline loader",
            },
            "children": Array [
              Object {
                "data": "admin.loading",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "style": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "style": undefined,
    },
    "x-attribsPrefix": Object {
      "style": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
